src/io/read_until.rs

```diff
@@ -0,0 +1,74 @@
+use std::io::{self, Read, BufRead};
+use std::mem;
+
+use futures::{Poll, Future};
+
+/// A future which can be used to easily read the contents of a stream into a 
+/// vector until the delimiter is reached.
+///
+/// Created by the [`read_until`] function.
+///
+/// [`read_until`]: fn.read_until.html
+pub struct ReadUntil<A> {
+    state: State<A>,
+}
+
+enum State<A> {
+    Reading {
+        a: A,
+        byte: u8,
+        buf: Vec<u8>,
+    },
+    Empty,
+}
+
+/// Creates a future which will read all the bytes associated with the I/O
+/// object `A` into the buffer provided until the delimiter `byte` is reached.
+/// This method is the async equivalent to [`BufRead::read_until`].
+///
+/// In case of an error the buffer and the object will be discarded, with
+/// the error yielded. In the case of success the object will be destroyed and
+/// the buffer will be returned, with all bytes up to, and including, the delimiter
+/// (if found).
+///
+/// [`ButRead::read_until`]: https://doc.rust-lang.org/std/io/trait.BufRead.html#method.read_until
+pub fn read_until<A>(a: A, byte: u8, buf: Vec<u8>) -> ReadUntil<A>
+    where A: BufRead
+{
+    ReadUntil {
+        state: State::Reading {
+            a: a,
+            byte: byte,
+            buf: buf,
+        }
+    }
+}
+
+impl<A> Future for ReadUntil<A>
+    where A: Read + BufRead
+{
+    type Item = (A, Vec<u8>);
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<(A, Vec<u8>), io::Error> {
+        match self.state {
+            State::Reading { ref mut a, byte, ref mut buf } => {
+                // If we get `Ok(n)`, then we know the stream hit EOF or the delimiter.
+                // If we got EOF (n == 0), then we return an error.
+                // If we hit "would block" then all the read data so far
+                // is in our buffer, and otherwise we propagate errors.
+                // If we reached our delimiter (n != 0), then we are finished.
+                let n = try_nb!(a.read_until(byte, buf));
+                if n == 0 {
+                    return Err(io::Error::new(io::ErrorKind::BrokenPipe, "Broken Pipe"));
```

Comment:
ah actually I think it needs to save off the size of `buf.len()` before the I/O even starts, because the vector could have data in it before we call `read_until`. (then the length after n ==0 is checked against the original length)